### ISSUE TITLE
Part of geometries missing at tiles borders for MVT

### ISSUE THREAD

**Issue by robrichter** [MapServer/mapcache#289]
We are serving MVT via Mapcache/Mapserver. Problem is that part of geometires at tiles borders are missing. We tried to use meta_buffer to solve the problem but its not implemented for raw format. Is there some workaround? 

![image](https://user-images.githubusercontent.com/8871336/172581675-98979471-25f3-4950-9659-703760eb170b.png)

**Comment by jratike80** [comment]
What software is doing the rendering part? It seems that the circle symbol in one tile is rendered correctly but then the adjacent tile is rendered on top of that layer and the the polygon fill paints partly over the symbol.

I do not know how these issues are usually handled on the client side vector tile rendering. The result should be good if all the polygon layers would be rendered first for all tiles, and the POI layer would be painted last. I need to read some vector tile style documents.

**Comment by jratike80** [comment]
Is that OpenLayers? We got a report about labels which were cut at the MVT tile boundaries. We noticed that an application that was using  MapLibre GL JS was showing labels fine but OpenLayers did not. We could correct the label issue by using an option `"declutter": true`. https://openlayers.org/en/latest/apidoc/module-ol_layer_VectorTile-VectorTileLayer.html. Maybe it could help also with symbols.

A possible lesson to learn is that because vector tiles are rendered on the client side then also the reason for the rendering issues may be in the client. Of course data can have an effect as well.

**Comment by robrichter** [comment]
Yes, we use OpenLayers. But we also tried [T-REX](https://t-rex.tileserver.ch/) as mvt source with exactly same Postgis data. We used  buffer_size=30. Everything worked fine. As I mentioned in my previous post, it seems to me that edge buffer in Mapserver is not working at all for us.  

We tried to use option "declutter": true on OpenLayers client and the results are strange. Some points were missing in lower zooms and started to appear when zooming in. And it also badly affected styling. Only some points were styled the rest remained blank, see screenshot below. 

![obrazek](https://user-images.githubusercontent.com/8871336/175301454-c0cb8e62-47de-4911-a44f-d70aed32ead6.png)

**Comment by jratike80** [comment]
You can open the .pbf tile files with QGIS as a plain vector layer (Layer-Add Layer-Add vector layer). Save the two adjacent .pbf tiles on disk and check if the problematic point is included only into one tile. If that is the case you are right with your assumption about edge buffer. You can by the same study if it happens to have some effect on lines and polygons.

**Comment by robrichter** [comment]
@sdlime Thanks
What do you need to prepare for the test case? I created github repo with mapcache config, mapfile and sql dump. Is that enough?

https://github.com/robrichter/mapcache-mvt-test

**Comment by sdlime** [comment]
> @sdlime Thanks What do you need to prepare for the test case? I created github repo with mapcache config, mapfile and sql dump. Is that enough?
> 
> https://github.com/robrichter/mapcache-mvt-test

Data is great. A simple client setup would also be helpful - a skeleton OpenLayers page.

**Comment by robrichter** [comment]
There is OL client example 
https://codesandbox.io/s/vector-tiles-openlayers-memj5x?file=/main.js

Please let me know if it is sufficient

**Comment by sdlime** [comment]
Thanks, will try to get that running on a dev box.

**Comment by theopauw** [comment]
@sdlime did you ever have a chance to look at this? I think I am running into the same underlying issue. 

I am using MapLibre and I find that points very close to the tile edge completely disappear at intermediate zoom levels.

<img width="450" alt="Image" src="https://github.com/user-attachments/assets/7f77d3b0-b107-4fa4-bd03-fc8456a7a1de" />  

<img width="450" alt="Image" src="https://github.com/user-attachments/assets/3fbdd5c4-478a-4b58-bb1a-7bcba2d775ef" />  

If I load the same service in QGIS the points are split in half similar to the original example.

<img width="450" height="349" alt="Image" src="https://github.com/user-attachments/assets/32f5b520-9e00-45ec-9b8d-a281796edd28" />

I tried patching Mapserver to apply the EDGE_BUFFER during feature selection as well (i.e. duplicating the point between the two adjacent tiles) and this fixes the issue for me.

<img width="450" alt="Image" src="https://github.com/user-attachments/assets/a79c0f1d-b03a-4288-8daf-8d512ed2e569" />  

<img width="450" alt="Image" src="https://github.com/user-attachments/assets/be3f1890-0f39-4c1f-8143-7adeaf905df3" />  

The fix works regardless of whether the request goes directly to Mapserver or through Mapcache.

See [my patched code](https://github.com/theopauw/MapServer/commit/d1022d494a7df0db4cc6738e557fbafd13365fed) and [example data, mapfile and maplibre html](https://github.com/theopauw/mapserver-maplibre-mvt-edge-issue). I can submit a PR to Mapserver if this behaviour is acceptable?
